src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -397,7 +399,22 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
         Set<String> roots = new HashSet<>();
         for (String mod : options.addMods) {
             if (mod.equals(ALL_MODULE_PATH) && options.modulePath.size() > 0) {
-                ModuleFinder mf = newModuleFinder(rootsFinder, options.limitMods, Set.of(), isLinkFromRuntime);
+                // Apply a module limit for the roots finder if it was otherwise empty.
+                // Since we are using the same finder for determining the roots set as
+                // for the actual link, we need to apply this trick so as to not
+                // include all modules of the JDK plus the extra module path in
+                // the roots set.
+                Set<String> allModsLimits = options.limitMods;
+                if (options.limitMods.isEmpty()) {
+                    Set<String> modsLimits = new HashSet<>();
+                    initialFinder.findAll()
+                                 .stream()
+                                 .map(ModuleReference::descriptor)
+                                 .map(ModuleDescriptor::name)
+                                 .forEach(mn -> modsLimits.add(mn));
+                    allModsLimits = modsLimits;
+                }
+                ModuleFinder mf = newModuleFinder(rootsFinder, allModsLimits, Set.of(), isLinkFromRuntime);
```

Comment:
It took me some time to understand this change as it depends on https://github.com/openjdk/jdk/pull/22609.   I adjusted the code trying to help the readers easier to understand.   

See what you think.   Instead of the `determineLinkFromRuntime` method, it does the check in this method that makes it explicit to the readers.

```
   private JlinkConfiguration initJlinkConfig() throws BadArgs {
        // Empty module path not allowed with ALL-MODULE-PATH in --add-modules
        if (options.addMods.contains(ALL_MODULE_PATH) && options.modulePath.isEmpty()) {
            throw taskHelper.newBadArgs("err.all.module.path.empty.mod.path");
        }

        ModuleFinder appModulePathFinder = createFinderFromPath(options.modulePath);
        ModuleFinder finder = appModulePathFinder;
        boolean isLinkFromRuntime = false;
        // if packaged module for java.base is not found, either from
        // the default module path or the run-time image
        if (!appModulePathFinder.find("java.base").isPresent()) {
            // Add the default module path if that exists
            Path defModPath = getDefaultModulePath();
            if (defModPath != null) {
                options.modulePath.add(defModPath);
                finder = createFinderFromPath(options.modulePath);
            }
            if (!finder.find("java.base").isPresent()) {
                isLinkFromRuntime = true;
                // java.base is from the system module path
                finder = ModuleFinder.compose(ModuleFinder.ofSystem(), appModulePathFinder);
            }
        }

        // Determine the roots set
        Set<String> roots = new HashSet<>();
        for (String mod : options.addMods) {
            if (mod.equals(ALL_MODULE_PATH)) {
                // all observable modules are roots
                Set<String> initialRoots = appModulePathFinder.findAll()
                        .stream()
                        .map(ModuleReference::descriptor)
                        .map(ModuleDescriptor::name)
                        .collect(Collectors.toSet());

                // Error if no module is found on module path
                if (initialRoots.isEmpty()) {
                    throw taskHelper.newBadArgs("err.all.module.path.empty.mod.path");
                }

                // Use a module finder with limited observability to find the observable modules
                // as not all JDK modules on the default module path or the run-time image
                ModuleFinder mf = newModuleFinder(finder, options.limitMods.isEmpty() ? initialRoots : options.limitMods,
                                                  Set.of(), isLinkFromRuntime);
                mf.findAll()
                  .stream()
                  .map(ModuleReference::descriptor)
                  .map(ModuleDescriptor::name)
                  .forEach(mn -> roots.add(mn));
            } else {
                roots.add(mod);
            }
        }
        finder = newModuleFinder(finder, options.limitMods, roots, isLinkFromRuntime);

        // --keep-packaged-modules doesn't make sense as we are not linking
        // from packaged modules to begin with.
        if (isLinkFromRuntime && options.packagedModulesPath != null) {
            throw taskHelper.newBadArgs("err.runtime.link.packaged.mods");
        }

        return new JlinkConfiguration(options.output,
                                      roots,
                                      finder,
                                      isLinkFromRuntime,
                                      options.ignoreModifiedRuntime,
                                      options.generateLinkableRuntime);
    }
```

Reply:
Is the `isLinkFromRuntime` parameter needed in the `newModuleFinder` method?   The java.base version check does not have any issue for linking from run-time image, right?

Reply:
> Is the `isLinkFromRuntime` parameter needed in the `newModuleFinder` method? The java.base version check does not have any issue for linking from run-time image, right?

It's not needed other than for skipping the check (as it'll trivially pass). I can remove it if that's preferred.